lib/codecs/src/decoding/format/protobuf.rs

```diff
@@ -0,0 +1,365 @@
+use std::collections::BTreeMap;
+use std::fs;
+
+use bytes::Bytes;
+use chrono::Utc;
+use lookup::PathPrefix;
+use ordered_float::NotNan;
+use prost_reflect::{DescriptorPool, DynamicMessage, MessageDescriptor, ReflectMessage};
+use smallvec::{smallvec, SmallVec};
+use vector_config::configurable_component;
+use vector_core::event::LogEvent;
+use vector_core::{
+    config::{log_schema, DataType, LogNamespace},
+    event::Event,
+    schema,
+};
+use vrl::value::Kind;
+
+use super::Deserializer;
+
+/// Config used to build a `ProtobufDeserializer`.
+#[configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct ProtobufDeserializerConfig {
+    /// Path to desc file
+    desc_file: String,
+
+    /// message type. e.g package.message
+    message_type: String,
+}
+
+impl ProtobufDeserializerConfig {
+    /// Build the `ProtobufDeserializer` from this configuration.
+    pub fn build(&self) -> ProtobufDeserializer {
+        ProtobufDeserializer::try_from(self).unwrap()
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                let mut definition =
+                    schema::Definition::empty_legacy_namespace().unknown_fields(Kind::any());
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    definition = definition.try_with_field(
+                        timestamp_key,
+                        // The protobuf decoder will try to insert a new `timestamp`-type value into the
+                        // "timestamp_key" field, but only if that field doesn't already exist.
+                        Kind::any().or_timestamp(),
+                        Some("timestamp"),
+                    );
+                }
+                definition
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::any(), [log_namespace])
+            }
+        }
+    }
+}
+
+impl ProtobufDeserializerConfig {
+    /// Creates a new `ProtobufDeserializerConfig`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing protobuf.
+#[derive(Debug, Clone)]
+pub struct ProtobufDeserializer {
+    message_descriptor: MessageDescriptor,
+}
+
+impl ProtobufDeserializer {
+    /// Creates a new `ProtobufDeserializer`.
+    pub fn new(message_descriptor: MessageDescriptor) -> Self {
+        Self { message_descriptor }
+    }
+
+    fn get_message_descriptor(
+        desc_file: String,
+        message_type: String,
```

Comment:
This function doesn't need to take ownership of either string, so these can both be just `&str`. Using an `&Path` for the filename would be even better (see comment above).